community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/SpatialLayoutTestUtil.java

```diff
@@ -0,0 +1,118 @@
+
+import org.apache.commons.lang3.ArrayUtils;
+
+import java.util.List;
+import java.util.Set;
+
+import org.neo4j.gis.spatial.index.curves.SpaceFillingCurve;
+import org.neo4j.index.internal.gbptree.Layout;
+import org.neo4j.internal.kernel.api.IndexQuery;
+import org.neo4j.kernel.api.index.IndexEntryUpdate;
+import org.neo4j.kernel.api.schema.index.SchemaIndexDescriptor;
+import org.neo4j.kernel.impl.index.schema.config.SpaceFillingCurveSettings;
+import org.neo4j.test.rule.RandomRule;
+import org.neo4j.values.storable.CoordinateReferenceSystem;
+import org.neo4j.values.storable.PointValue;
+import org.neo4j.values.storable.Value;
+import org.neo4j.values.storable.Values;
+
+import static org.neo4j.values.storable.CoordinateReferenceSystem.WGS84;
+
+class SpatialLayoutTestUtil extends LayoutTestUtil<SpatialSchemaKey,NativeSchemaValue>
+{
+    private static final PointValue[] ALL_EXTREME_VALUES = new PointValue[]
+    {
+            Values.pointValue( WGS84, -180, -90 ),
+            Values.pointValue( WGS84, -180, 90 ),
+            Values.pointValue( WGS84, 0, 0 ),
+            Values.pointValue( WGS84, -1, 1 ),
+            Values.pointValue( WGS84, 180, 90 ),
+            Values.pointValue( WGS84, 180, -90 ),
+    };
+
+    private final CoordinateReferenceSystem crs;
+    private final SpaceFillingCurve curve;
+
+    SpatialLayoutTestUtil( SchemaIndexDescriptor descriptor, SpaceFillingCurveSettings settings, CoordinateReferenceSystem crs )
+    {
+        super( descriptor );
+        this.curve = settings.curve();
+        this.crs = crs;
+        // The layout is the same, but we might consider supporting other CRS here, too.
+        assert crs == CoordinateReferenceSystem.WGS84;
+    }
+
+    @Override
+    Layout<SpatialSchemaKey,NativeSchemaValue> createLayout()
+    {
+        return new SpatialLayout( crs, curve );
+    }
+
+    @Override
+    IndexEntryUpdate<SchemaIndexDescriptor>[] someUpdates()
+    {
+        return someUpdatesWithDuplicateValues();
+    }
+
+    @Override
+    IndexQuery rangeQuery( Object from, boolean fromInclusive, Object to, boolean toInclusive )
+    {
+        return IndexQuery.range( 0, (PointValue) from , fromInclusive, (PointValue) to, toInclusive );
+    }
+
+    @Override
+    int compareIndexedPropertyValue( SpatialSchemaKey key1, SpatialSchemaKey key2 )
+    {
+        return Long.compare( key1.rawValueBits, key2.rawValueBits );
+    }
+
+    @Override
+    Value newUniqueValue( RandomRule random, Set<Object> uniqueCompareValues, List<Value> uniqueValues )
+    {
+        double x, y;
+        PointValue pointValue;
+        Long compareValue;
+        do
+        {
+            x = random.nextDouble() * 360 - 180;
+            y = random.nextDouble() * 180 - 90;
+            pointValue = Values.pointValue( crs, x, y );
+            compareValue = curve.derivedValueFor( pointValue.coordinate() );
+        }
+        while ( !uniqueCompareValues.add( compareValue ) );
```

Comment:
You're right. Unfortunately the test framework only exercises that part of the code, without the subsequent filtering that is needed for Spatial. That means if we allow Points that map to the same 1D-value here, the tests are failing because they do not expect to get more than one result for a query, but get more. The framework is not really designed for lossy indexes, but still provides _some_ nice coverage for Spatial.